__tests__/unit/cli/test_utils.py

```diff
@@ -0,0 +1,128 @@
+            file.parent.mkdir(parents=True, exist_ok=True)
+            file.touch()
+
+        # Act
+        zipdir(tmp_path, mock_zip_archive)
+
+        # Assert
+        for full_path, relative_path in all_files.items():
+            mock_zip_archive.write.assert_any_call(str(full_path), relative_path)
+
+    @staticmethod
+    def test_excluded_folders(tmp_path: pathlib.Path, mocker) -> None:
+        """Assert that a top-level folder and its children is excluded."""
+        # Arrange
+        mock_zip_archive = mocker.MagicMock()
+        # fmt: off
+        all_files = {
+            tmp_path.joinpath("foo", "January", "red", "file.py"): "foo/January/red/file.py",
+            tmp_path.joinpath("foo", "January", "green", "file.py"): "foo/January/green/file.py",
+            tmp_path.joinpath("foo", "February", "red", "file.py"): "foo/February/red/file.py",
+            tmp_path.joinpath("foo", "February", "green", "file.py"): "foo/February/green/file.py",
+            tmp_path.joinpath("foo", "March", "red", "file.py"): "foo/March/red/file.py",
+            tmp_path.joinpath("foo", "March", "green", "file.py"): "foo/March/green/file.py",
+            tmp_path.joinpath("bar", "January", "red", "file.py"): "bar/January/red/file.py",
+            tmp_path.joinpath("bar", "January", "green", "file.py"): "bar/January/green/file.py",
+            tmp_path.joinpath("bar", "February", "red", "file.py"): "bar/February/red/file.py",
+            tmp_path.joinpath("bar", "February", "green", "file.py"): "bar/February/green/file.py",
+            tmp_path.joinpath("bar", "March", "red", "file.py"): "bar/March/red/file.py",
+            tmp_path.joinpath("bar", "March", "green", "file.py"): "bar/March/green/file.py",
+        }
+        excluded_folders = ["foo"]
+        expected_written_files = {
+            tmp_path.joinpath("bar", "January", "red", "file.py"): "bar/January/red/file.py",
+            tmp_path.joinpath("bar", "January", "green", "file.py"): "bar/January/green/file.py",
+            tmp_path.joinpath("bar", "February", "red", "file.py"): "bar/February/red/file.py",
+            tmp_path.joinpath("bar", "February", "green", "file.py"): "bar/February/green/file.py",
+            tmp_path.joinpath("bar", "March", "red", "file.py"): "bar/March/red/file.py",
+            tmp_path.joinpath("bar", "March", "green", "file.py"): "bar/March/green/file.py",
+        }
+        # fmt: on
+        for file in all_files:
+            file.parent.mkdir(parents=True, exist_ok=True)
+            file.touch(exist_ok=True)
+
+        # Act
+        zipdir(tmp_path, mock_zip_archive, folders_to_exclude=excluded_folders)
+
+        # Assert
+        for full_path, relative_path in expected_written_files.items():
+            mock_zip_archive.write.assert_any_call(str(full_path), relative_path)
+        # Assert that `write` was not called with any file other than those asserted against above.
+        assert mock_zip_archive.write.call_count == 6
```

Comment:
It would be nice if this were:
```suggestion
        assert mock_zip_archive.write.call_count == len(expected_written_files)
```
This saves me from having to count the length of `expected_written_files` to confirm it matches.  I'm not sure if the comment above would still be necessary, but it's hard for me to tell.

Reply:
IMO that's too much logic for a unit test. I still implemented it.